### ISSUE TITLE
Reduce IPv6 default lifetimes

### ISSUE THREAD

**Issue by jpds** [GrapheneOS/os-issue-tracker#235]
The default lifetimes for IPv6 privacy extensions should be potentially reduced to one hour and three hours. The effect of this can be tested with a much shorter lifetime of:

```
sudo sysctl -w net.ipv6.conf.wls7.use_tempaddr=2
sudo sysctl -w net.ipv6.conf.wls7.temp_prefered_lft=250
sudo sysctl -w net.ipv6.conf.wls7.temp_valid_lft=300
```

And then:

```
watch ip -6 a s
```

Once your preferred lifetime hits 0, Linux will request a new temporary address and mark the old one deprecated (allowing time for connections to close) and then when the valid_lft hits zero, it's removed from the kernel.

**Comment by thestinger** [comment]
What happens if we set it to 0?

**Comment by jpds** [comment]
If I set both to `0`, it uses the lifetimes configured in the router advertisements - however it resets the counters when a new advertisement is received and thus the address never changes.

**Comment by jpds** [comment]
https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt

Could be worth setting `addr_gen_mode` to `3`.

**Comment by jpds** [comment]
Going back to the addresses from here: https://github.com/GrapheneOS/os_issue_tracker/issues/223#issuecomment-648937135

There is definitely something (such as a stable secret) on the device that's setting that `:3faf` address - I just went back to the other IPv6 network I've been testing this and that same device network ending is on there (I do not administer that network at all and it's just a standard café WiFi).

**Comment by thestinger** [comment]
Disabling privacy addresses is what would actually fix the problem, when MAC randomization is enabled, but that would be a problem when MAC randomization isn't enabled. This is an issue with IPv6 and the Linux implementation of it. I don't think AOSP is doing anything wrong in particular.

**Comment by thestinger** [comment]
The stable secret feature is not being used anymore.

**Comment by thestinger** [comment]
What happens if you reboot the device? There should be no state preserved between reboots since the stable secret feature is not used. It's entirely the kernel's implementation of privacy addresses. The Linux kernel just doesn't work in a reasonable way.

**Comment by jpds** [comment]
> What happens if you reboot the device?

The device comes back with the same "stable" address.

I've just gone back to the same café and when I got there my "stable" address had rotated, now ending in `da84:2dfe`, the MAC here ends in `61:19`:

![signal-2020-06-26-105220_001](https://user-images.githubusercontent.com/29158971/85845169-c28c2600-b79b-11ea-92ed-0887d4b08e8f.jpeg)

I then went back home and got these addresses, MAC ending in `ee:71`:

![signal-2020-06-26-105220_002](https://user-images.githubusercontent.com/29158971/85845242-dafc4080-b79b-11ea-9a76-c5c793458273.jpeg)

But the `da84:2dfe` persists across the two independent networks, but this isn't the address I had before so it would appear that there's a mechanism somewhere in Android that decides to rotate these "stable" device addresses after a certain number of days. It would be perfect if this could be triggered before a WiFi point is connected to.

**Comment by thestinger** [comment]
Yeah so this is how the Linux kernel does "privacy" addresses. This privacy issue is a consequence of having the "privacy" address feature enabled. If it was disabled, this would just be the (random) MAC based address. With full MAC randomization enabled, that would offer better privacy.

**Comment by jpds** [comment]
Yep, I understand, my concern with just the MAC approach would be someone connected to the same WiFi for days and having the same v6 address for that specific device that's global on the internet. It's different to v4 where you would be behind a NAT/CGNAT with multiple devices.

Linux _should_ be leveraging the temporary addresses with limited lifetimes and rotating those addresses over hours or days.

**Comment by thestinger** [comment]
It does appear to do that (1 day soft limit, 7 day hard limit by default) but the issue is it's not per-network...

**Comment by thestinger** [comment]
And lowering that time won't fix that it's not per-network. That's a problem even if it's something like 5 minutes, and I'm pretty sure the hard limit will act as a limit on how long connections can last.

**Comment by jpds** [comment]
At the current time, it probably is better to remove the privacy addresses for now.

**Comment by thestinger** [comment]
Replacing with https://github.com/GrapheneOS/os-issue-tracker/issues/557. I don't think we want to change the lifetime while connected but rather just fix them by flushing the state properly between connections to networks.

**Comment by thestinger** [comment]
@jpds The https://github.com/GrapheneOS/os-issue-tracker/issues/557 issue will be fixed in the next release.

There are still some more subtle issues enabling tracking in the Linux kernel but at least this one will be out of the way.

**Comment by thestinger** [comment]
https://github.com/torvalds/linux/commit/969c54646af0d7d94a5f0f37adbbfe024e85466e is the upstream fix for Linux 5.8 and later reimplementing privacy addresses in a way that actually provides privacy instead of enabling tracking. The issue was that this isn't present for anything but the latest 5.10 LTS branch. We're applying a much simpler, less sophisticated workaround for existing devices and we shouldn't need to change the upstream implementation once devices are on the most recent 5.10 LTS branch.
